### PR TITLE
fix!: make negotiation and reconnect state transactional

### LINKED ISSUE #100: Track requested and effective game-data encoding separately
## Problem

The core retains the configured/requested `GameDataEncoding` and ignores `ProtocolInfo.game_data_formats`. It downgrades to JSON only after an `UnsupportedGameDataFormat` error. An Rkyv-configured client can therefore send an unsupported binary frame and quarantine legitimate JSON data before or without that error.

Found during #97.

## Invariant

Physical frame validation and outbound encoding must use the server-negotiated effective format, while preserving the caller's requested preference for diagnostics/reconnect.

## Acceptance criteria

- Model requested and effective formats separately in shared state.
- Resolve the effective format atomically from the first authoritative ProtocolInfo using the pinned Server 0.7 negotiation rules.
- Prevent unsupported outbound binary frames before transport admission.
- Validate inbound JSON/binary frames against the effective format.
- Expose the effective format through a coherent public snapshot/accessor.
- Add async/polling parity tests for supported preference, Rkyv-to-JSON fallback, malformed/duplicate negotiation, reconnect, and messages arriving around negotiation.
- Add blocking real-server evidence for at least one fallback path.

### LINKED ISSUE #101: Make reconnect negotiation, accountability, and snapshot restoration atomic
## Problem

The client defensively accepts replayed `ProtocolInfo` and `SessionPlan` entries even though the pinned Server 0.7 reconnect shape excludes them. A replayed ProtocolInfo can restore snapshot protocol v3 while delivery accountability remains in v2 mode. V3 reconnect payload validation also permits incomplete snapshots (including empty membership/watermarks and missing replay/token fields), and intended WebRTC continuity before the replacement live SessionPlan is unspecified.

Found during #97.

## Invariant

A successful reconnect must restore one internally consistent version/accountability/membership/plan state or fail without partial mutation.

## Acceptance criteria

- Decide and document whether noncanonical replayed ProtocolInfo/SessionPlan are rejected or supported defensive extensions.
- If supported, determine effective version before validating the reconnect and switch accountability atomically.
- Validate version-sensitive reconnect payload invariants: local member exactly once, complete/nonempty watermarks where required, replay status, and rotated token; apply analogous self-membership checks to RoomJoined.
- Confirm with the server whether an existing WebRTC data channel remains authoritative before a fresh live SessionPlan; fence or explicitly preserve driver output accordingly.
- Add shared-core plus async/polling tests for valid v2/v3 reconnects, malformed snapshots, replayed negotiation, stale plan output, and rollback/no-partial-mutation behavior.
- Add real-server reconnect evidence.

### PR DESCRIPTION
**Pull request by wallstop**
## Summary

- separate the requested game-data preference from the effective Server 0.7-negotiated format across the shared core, snapshots, async/polling clients, and object-safe API
- validate and commit the first canonical `ProtocolInfo` atomically; enforce JSON fallback before transport admission while preserving JSON-origin text relays for MessagePack recipients
- make reconnect terminal responses causally match an admitted player/room/token request, validate v2/v3 replay-accountability baselines transactionally under every policy, and fence the old WebRTC plan until the server's fresh live plan
- add async/polling parity matrices, malformed-state rollback tests, pinned Server 0.7 acceptance tests, blocking CI rows, public docs, changelog, PLAN, and session evidence

## Compatibility

This adds two fields to the exhaustive `ClientSnapshot` struct and is therefore recorded as a breaking public-API change for the forthcoming 0.11 release. `SignalFishClientApi` supplies default accessors through `snapshot()`.

## Verification

- `cargo fmt`
- `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- `cargo test --workspace --all-features`
- no-default, polling-only, and Tokio-only feature checks
- rustdoc with warnings denied
- local docs validation
- all 18 pre-commit and all 6 pre-push checks
- pinned Server 0.7 ARM64 binary:
  - `e2e_server_070_rkyv_request_resolves_to_json`
  - `e2e_reconnect_after_disconnect_uses_server_token`
- three independent audits plus adversarial re-review; final review reported no remaining findings

Closes #100.
Closes #101.

---

> [!NOTE]
> **High Risk**
> Touches shared protocol state, reconnect credentials, mesh/WebRTC fencing, and breaking public snapshot APIs—incorrect validation or barrier ordering could break live sessions or mis-route signals.
> 
> **Overview**
> **Game-data negotiation** now keeps **requested** vs **effective** formats separate on `ClientSnapshot` and async/polling/`SignalFishClientApi` accessors. The shared core resolves the first canonical Server 0.7 `ProtocolInfo` atomically (unsupported or omitted requests → JSON); `UnsupportedGameDataFormat` no longer mutates negotiated state. Binary admission and frame validation use the effective format, while **JSON-origin text relays stay valid** for MessagePack peers.
> 
> **Reconnect** is stricter and causally bound: terminal `Reconnected` / `ReconnectionFailed` must follow an admitted `Reconnect`; v3 baselines require replay, rotated token, stamps, and complete watermarks, and invalid baselines never apply (including under `Observe`). **`Reconnected` fences the WebRTC plan**—`MeshSession` and `MeshController` drop the old plan/peers until a fresh live `SessionPlan`; the mesh controller advances plan barriers per consumed event to avoid coalesced replan races.
> 
> Docs, changelog, CI e2e rows (Rkyv→JSON, reconnect token), and broad parity/core/mesh tests cover the above; `ClientSnapshot` gains two fields (breaking for 0.11).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b61b173247f191e84ff9580cb63c7245bb974554. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/webrtc.rs [hunk 1]
```diff
@@ -371,9 +385,19 @@ mod controller {
...
+            if matches!(
+                event,
+                SignalFishEvent::RoomJoined { .. }
+                    | SignalFishEvent::Reconnected { .. }
+                    | SignalFishEvent::SessionPlan { .. }
+                    | SignalFishEvent::RoomLeft
+                    | SignalFishEvent::Disconnected { .. }
+            ) {
+                self.observed_session_plan_revision = self.client.session_plan_revision();
+            }
```
[the remark this answers is not in the document]
Comment (wallstop) [reply src/webrtc.rs:411]: Fixed in 90ef991. `MeshController` now advances exactly one revision for each room/plan/spectator/disconnect barrier it actually consumes, instead of copying the core's potentially-ahead counter. Spectator transitions also reset authoritative mesh state and tear down known peers, and core session clear now creates a matching barrier. The regression queues two plans, advances the core to the second while the controller still holds the first, injects old-generation Data, and proves the second SessionPlan is folded before any stale driver output can surface. The full workspace/all-feature gate and targeted WebRTC/mesh tests pass locally.

File: src/client_core.rs [hunk 2]
```diff
@@ -438,6 +472,9 @@ impl ClientCore {
         self.session_peers.clear();
         self.session_transport = None;
         self.retired_generationless_signal_peers.clear();
+        self.pending_reconnects.clear();
+        #[cfg(feature = "tokio-runtime")]
+        self.advance_room_revision();
```
[the remark this answers is not in the document]
Comment (wallstop) [reply src/client_core.rs]: Fixed in b61b173. `clear_session` now advances only `session_plan_revision`; `room_revision` remains reserved for actual room-membership changes. Added `reliable_game_data_reports_disconnect_after_waiting_for_capacity`, which blocks a reliable room-scoped send on queue capacity, disconnects the core, releases capacity, and asserts the result is `NotConnected` (not `NotInRoom`). The full mandatory workspace gate plus all pre-commit/pre-push checks pass.
